File: ros2-for-beginners/ros2_ws/src/my_py_pkg/my_py_pkg/add_two_ints_server.py
#!/usr/bin/python3

import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts


class AddTwoIntNode(Node):
    def __init__(self) -> None:
        super().__init__("add_two_ints_server")

        """
        create a service server instance 
            - using create_service method in Node class
                --> takes in 3 arguements
                    1. service message type
                    2. service name
                    3. callback function - for handling the request to the server 

        """
        self.server_ = self.create_service(
            AddTwoInts, "add_two_ints", self.callback_add_two_ints)

        self.get_logger().info("<-- python add ints server started -->")

    """
    call back function for handling the request and responding to it 
    it takes 2 parameters 
        1. request 
        2. response 
    """

    def callback_add_two_ints(self, request, response):
        response.sum = request.a + request.b
        self.get_logger().info(f"{request.a} + {request.b} = {response.sum}")
        return response


def main(args=None):
    rclpy.init()
    node = AddTwoIntNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: ros2-for-beginners/ros2_ws/src/my_cpp_pkg/src/add_two_ints_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp" 

using std::placeholders::_1;
using std::placeholders::_2;

class AddTwoIntsServerNode : public rclcpp::Node{
    public:
        AddTwoIntsServerNode() : Node("add_two_ints_server") {

            server_ = this->create_service<example_interfaces::srv::AddTwoInts>("add_two_ints",
            std::bind(&AddTwoIntsServerNode::callbackAddTwoInts,this,_1,_2));

            RCLCPP_INFO(this->get_logger(),"cpp add two ints server started");
        }

    private:

    // callback
    // params -> req, res
    void callbackAddTwoInts(const example_interfaces::srv::AddTwoInts::Request::SharedPtr request,
    const example_interfaces::srv::AddTwoInts::Response::SharedPtr response){
            // compute sum
            response->sum = request->a + request->b;
            RCLCPP_INFO(this->get_logger(),"%ld + %ld = %ld",request->a,request->b,response->sum);
    }

    // make a server variable - declaration
    rclcpp::Service<example_interfaces::srv::AddTwoInts>::SharedPtr server_; 


};


int main(int argc, char **argv){
    rclcpp::init(argc,argv);
    auto node = std::make_shared<AddTwoIntsServerNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}